File: Final_Project/Code/Project_4/Bai_toan_5/Problems/3/CSS_bt5_problem3.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<pair<int,int>>> all_trees;

struct Graph {
    int n;
    vector<vector<int>> adj;
    Graph(int _n): n(_n), adj(n, vector<int>(n,0)) {}
    void add_edge(int u, int v) {
        // Them canh (u,v)
        adj[u][v] = adj[v][u] = 1;
    }
    vector<pair<int,int>> edges() const {
        // Tra ve cac cap (u,v) voi u < v
        vector<pair<int,int>> E;
        for(int u = 0; u < n; ++u)
            for(int v = u+1; v < n; ++v)
                if(adj[u][v])
                    E.emplace_back(u, v);
        return E;
    }
};

struct UF {
    vector<int> parent;
    int count;
    UF(int n): parent(n), count(n) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int x) {
        return parent[x] == x ? x : parent[x] = find(parent[x]);
    }
    bool unite(int a, int b) {
        a = find(a); b = find(b);
        if(a == b) return false;
        parent[b] = a;
        --count;
        return true;
    }
};

void backtrack(const vector<pair<int,int>>& E,
               int n, int idx,
               vector<pair<int,int>>& chosen) {
    if((int)chosen.size() == n - 1) {
        UF uf(n);
        for(auto &e : chosen)
            if(!uf.unite(e.first, e.second))
                return;
        if(uf.count == 1)
            all_trees.push_back(chosen);
        return;
    }
    if(idx == (int)E.size()) return;
    // Bo chon E[idx]
    backtrack(E, n, idx + 1, chosen);
    // Chon E[idx]
    chosen.push_back(E[idx]);
    backtrack(E, n, idx + 1, chosen);
    chosen.pop_back();
}

int main() {
    Graph G(7);
    vector<pair<int,int>> edges = {
        {1,2},{1,3},{3,6},{6,4},{4,1},
        {2,4},{4,7},{7,2},{2,5},{5,7},{6,7}
    };
    for(auto &e : edges)
        G.add_edge(e.first-1, e.second-1);
    auto E = G.edges();
    vector<pair<int,int>> chosen;
    backtrack(E, G.n, 0, chosen);
    cout << "So spanning trees tim duoc: "
         << all_trees.size() << "\n";
    return 0;
}

File: Final_Project/Code/Project_4/Bai_toan_5/Problems/3/CSS_bt5_problem3.py
class Graph:
    def __init__(self, n):
        self.n = n
        self.adj_matrix = [[0]*n for _ in range(n)]
    def add_edge(self, u, v):
        # Them canh (u,v)
        self.adj_matrix[u][v] = 1
        self.adj_matrix[v][u] = 1
    def edges(self):
        # Tra ve cac cap (u,v) voi u < v
        return [(u, v)
                for u in range(self.n)
                for v in range(u+1, self.n)
                if self.adj_matrix[u][v]]

class UF:
    def __init__(self, n):
        self.parent = list(range(n))
        self.count = n
    def find(self, u):
        # Path compression
        if self.parent[u] != u:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]
    def union(self, u, v):
        ru, rv = self.find(u), self.find(v)
        if ru == rv:
            return False
        self.parent[rv] = ru
        self.count -= 1
        return True

def enumerate_spanning_trees(G):
    n = G.n
    E = G.edges()
    all_trees = []
    chosen = []

    def backtrack(idx):
        if len(chosen) == n - 1:
            uf = UF(n)
            for u, v in chosen:
                if not uf.union(u, v):
                    return
            if uf.count == 1:
                all_trees.append(chosen.copy())
            return
        if idx == len(E):
            return
        # Bo chon E[idx]
        backtrack(idx + 1)
        # Chon E[idx]
        chosen.append(E[idx])
        backtrack(idx + 1)
        chosen.pop()

    backtrack(0)
    return all_trees

if __name__ == "__main__":
    G = Graph(7)
    # Chuyen 1-based -> 0-based
    for u, v in [(1,2),(1,3),(3,6),(6,4),(4,1),
                 (2,4),(4,7),(7,2),(2,5),(5,7),(6,7)]:
        G.add_edge(u-1, v-1)
    trees = enumerate_spanning_trees(G)
    print("So spanning trees tim duoc:", len(trees))